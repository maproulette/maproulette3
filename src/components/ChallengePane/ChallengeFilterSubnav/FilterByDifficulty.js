import React, { Component } from 'react'
import PropTypes from 'prop-types'
import _map from 'lodash/map'
import { injectIntl } from 'react-intl'
import NavDropdown from '../../Bulma/NavDropdown'
import { ChallengeDifficulty,
         difficultyLabels }
       from '../../../services/Challenge/ChallengeDifficulty/ChallengeDifficulty'
import messages from './Messages'

/**
 * FilterByDifficulty displays a nav dropdown containing options for filtering
 * challenges by difficulty or suggested experience. The challenge filter in
 * the redux store is updated to reflect the selected difficulty option.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class FilterByDifficulty extends Component {
  /**
   * Update the challenge filter with the selected value.
   *
   * @private
   */
  updateFilter = ({ value }) => {
    if (value === null) {
      this.props.removeChallengeFilters(['difficulty'])
    }
    else {
      this.props.setChallengeFilters({difficulty: value})
    }
  }

  render() {
    const localizedDifficultyLabels = difficultyLabels(this.props.intl)

    const selectOptions = _map(ChallengeDifficulty, (difficulty, name) => ({
      key: difficulty,
      text: localizedDifficultyLabels[name],
      value: difficulty,
    }))

    // Add 'Any' option to start of dropdown
    const anyOption = {
      key: 'any',
      text: localizedDifficultyLabels.any,
      value: null,
    }
    selectOptions.unshift(anyOption)

    return (
      <NavDropdown placeholder={anyOption.text}
                   label={this.props.intl.formatMessage(messages.difficultyLabel)}
                   options={selectOptions}
                   value={this.props.challengeFilter.difficulty}
                   onChange={this.updateFilter}
      />
    )
  }
}

FilterByDifficulty.propTypes = {
  /** Invoked to update the challenge difficulty filter */
  setChallengeFilters: PropTypes.func.isRequired,
  /** Invoked to clear the challenge difficulty filter */
  removeChallengeFilters: PropTypes.func.isRequired,
  /** The current value of the challenge filter */
  challengeFilter: PropTypes.object,
}

FilterByDifficulty.defaultProps = {
  challengeFilter: {},
}

export default injectIntl(FilterByDifficulty)
