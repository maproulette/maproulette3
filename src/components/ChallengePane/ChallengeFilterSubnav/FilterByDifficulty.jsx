import _map from "lodash/map";
import PropTypes from "prop-types";
import { Component } from "react";
import { FormattedMessage, injectIntl } from "react-intl";
import {
  ChallengeDifficulty,
  difficultyLabels,
  messagesByDifficulty,
} from "../../../services/Challenge/ChallengeDifficulty/ChallengeDifficulty";
import Dropdown from "../../Dropdown/Dropdown";
import ButtonFilter from "./ButtonFilter";
import messages from "./Messages";

/**
 * FilterByDifficulty displays a nav dropdown containing options for filtering
 * challenges by difficulty or suggested experience. The challenge filter in
 * the redux store is updated to reflect the selected difficulty option
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class FilterByDifficulty extends Component {
  /**
   * Update the challenge filter with the selected value.
   *
   * @private
   */
  updateFilter = (value, closeDropdown) => {
    if (!Number.isFinite(value)) {
      this.props.removeSearchFilters(["difficulty"]);
    } else {
      this.props.setSearchFilters({ difficulty: value });
    }
    closeDropdown();
  };

  render() {
    const localizedDifficultyLabels = difficultyLabels(this.props.intl);
    const notFiltering = !Number.isFinite(this.props.searchFilters.difficulty);

    return (
      <Dropdown
        className="mr-dropdown--flush xl:mr-border-l xl:mr-border-white-10 mr-p-6 mr-pl-0 xl:mr-pl-6"
        dropdownButton={(dropdown) => (
          <ButtonFilter
            type={<FormattedMessage {...messages.difficultyLabel} />}
            selection={
              notFiltering ? (
                localizedDifficultyLabels.any
              ) : (
                <FormattedMessage {...messagesByDifficulty[this.props.searchFilters.difficulty]} />
              )
            }
            selectionClassName={notFiltering ? null : "mr-text-yellow"}
            onClick={dropdown.toggleDropdownVisible}
          />
        )}
        dropdownContent={(dropdown) => (
          <ListDifficultyTypes
            difficultyLabels={localizedDifficultyLabels}
            updateFilter={this.updateFilter}
            closeDropdown={dropdown.closeDropdown}
          />
        )}
      />
    );
  }
}

const ListDifficultyTypes = function (props) {
  const menuItems = _map(ChallengeDifficulty, (difficulty, name) => (
    <li key={difficulty}>
      <a onClick={() => props.updateFilter(difficulty, props.closeDropdown)}>
        {props.difficultyLabels[name]}
      </a>
    </li>
  ));

  // Add 'Any' option to start of dropdown
  menuItems.unshift(
    <li key="any">
      <a onClick={() => props.updateFilter(null, props.closeDropdown)}>
        {props.difficultyLabels.any}
      </a>
    </li>,
  );

  return <ol className="mr-list-dropdown mr-list-dropdown--ruled">{menuItems}</ol>;
};

FilterByDifficulty.propTypes = {
  /** Invoked to update the challenge difficulty filter */
  setSearchFilters: PropTypes.func.isRequired,
  /** Invoked to clear the challenge difficulty filter */
  removeSearchFilters: PropTypes.func.isRequired,
  /** The current value of the challenge filter */
  searchFilters: PropTypes.object,
  /** Set to true to render as a list instead of a dropdown */
  asMenuList: PropTypes.bool,
};

FilterByDifficulty.defaultProps = {
  searchFilters: {},
};

export default injectIntl(FilterByDifficulty);
