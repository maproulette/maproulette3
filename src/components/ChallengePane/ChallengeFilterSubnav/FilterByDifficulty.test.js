import React from 'react'
import { FilterByDifficulty } from './FilterByDifficulty'
import _cloneDeep from 'lodash/cloneDeep'

const propsFixture = {
  user: {
    id: 11,
    savedChallenges: [],
  },
  challenges: [
    {
      id: 309,
      name: "Challenge 309",
      blurb: "Challenge 309 blurb",
      description: "Challenge 309 description",
      parent: {
        displayName: "foo",
      }
    },
    {
      id: 311,
      name: "Challenge 311",
      blurb: "Challenge 311 blurb",
      description: "Challenge 311 description",
      parent: {
        displayName: "bar",
      }
    }
  ],
  challengeFilter: {
    difficulty: 1
  }
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)
  basicProps.setChallengeFilters = jest.fn()
  basicProps.removeChallengeFilters = jest.fn()
  basicProps.intl = {formatMessage: jest.fn()}
})

test("it renders with props as expected", () => {
  const wrapper = shallow(
    <FilterByDifficulty {...basicProps} />
  )

  expect(wrapper).toMatchSnapshot()
})
