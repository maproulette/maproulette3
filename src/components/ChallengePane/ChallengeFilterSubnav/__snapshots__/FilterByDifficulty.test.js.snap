// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with props as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterByDifficulty
    asMenuList={false}
    challenges={
      Array [
        Object {
          "blurb": "Challenge 309 blurb",
          "description": "Challenge 309 description",
          "id": 309,
          "name": "Challenge 309",
          "parent": Object {
            "displayName": "foo",
          },
        },
        Object {
          "blurb": "Challenge 311 blurb",
          "description": "Challenge 311 description",
          "id": 311,
          "name": "Challenge 311",
          "parent": Object {
            "displayName": "bar",
          },
        },
      ]
    }
    intl={
      Object {
        "formatMessage": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "defaultMessage": "Easy",
                "id": "Challenge.difficulty.easy",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Normal",
                "id": "Challenge.difficulty.normal",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Expert",
                "id": "Challenge.difficulty.expert",
              },
            ],
            Array [
              Object {
                "defaultMessage": "Any",
                "id": "Challenge.difficulty.any",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "Easy",
            },
            Object {
              "isThrow": false,
              "value": "Normal",
            },
            Object {
              "isThrow": false,
              "value": "Expert",
            },
            Object {
              "isThrow": false,
              "value": "Any",
            },
          ],
        },
      }
    }
    removeSearchFilters={[MockFunction]}
    searchFilters={
      Object {
        "difficulty": 1,
      }
    }
    setSearchFilters={[MockFunction]}
    user={
      Object {
        "id": 11,
        "savedChallenges": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "eventTypes": Array [
        "mousedown",
        "touchstart",
      ],
      "excludeScrollbar": false,
      "label": <FormattedMessage
        defaultMessage="Difficulty"
        id="ChallengeFilterSubnav.filter.difficulty.label"
        values={Object {}}
      />,
      "onChange": [Function],
      "options": Array [
        Object {
          "key": "any",
          "text": "Any",
          "value": undefined,
        },
        Object {
          "key": 1,
          "text": "Easy",
          "value": 1,
        },
        Object {
          "key": 2,
          "text": "Normal",
          "value": 2,
        },
        Object {
          "key": 3,
          "text": "Expert",
          "value": 3,
        },
      ],
      "outsideClickIgnoreClass": "ignore-react-onclickoutside",
      "placeholder": "Any",
      "preventDefault": false,
      "stopPropagation": false,
      "value": 1,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "eventTypes": Array [
          "mousedown",
          "touchstart",
        ],
        "excludeScrollbar": false,
        "label": <FormattedMessage
          defaultMessage="Difficulty"
          id="ChallengeFilterSubnav.filter.difficulty.label"
          values={Object {}}
        />,
        "onChange": [Function],
        "options": Array [
          Object {
            "key": "any",
            "text": "Any",
            "value": undefined,
          },
          Object {
            "key": 1,
            "text": "Easy",
            "value": 1,
          },
          Object {
            "key": 2,
            "text": "Normal",
            "value": 2,
          },
          Object {
            "key": 3,
            "text": "Expert",
            "value": 3,
          },
        ],
        "outsideClickIgnoreClass": "ignore-react-onclickoutside",
        "placeholder": "Any",
        "preventDefault": false,
        "stopPropagation": false,
        "value": 1,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
